File: Programmers/cpp/LEVEL_3_불량_사용자.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_set>
using namespace std;

vector<vector<string>> list;
vector<string> strList;

void getList(int cnt, int limit, vector<string>& userId, vector<bool>& visit) {
	if (cnt == limit) {
		vector<string> st(strList.size());
		for (int i = 0; i < strList.size(); i++) {
			st[i] = strList[i];
		}

		list.push_back(st);

		return;
	}
	else {
		for (int i = 0; i < userId.size(); i++) {
			if (visit[i] == true) continue;
			visit[i] = true;
			strList.push_back(userId[i]);
			getList(cnt + 1, limit, userId, visit);
			visit[i] = false;
			strList.pop_back();
		}
	}
}

bool validateWord(string word, string bannedWord) {
	if (word.length() != bannedWord.length()) {
		return false;
	}

	for (int i = 0; i < word.length(); i++) {
		if (bannedWord[i] == '*') {
			continue;
		}

		if (word[i] != bannedWord[i]) {
			return false;
		}
	}

	return true;
}

vector<vector<string>> searchBannedId(vector<vector<string>>& list, vector<string>& banned_id) {
	vector<vector<string>> ret;

	for (vector<string> str : list) {
		bool flag = true;
		for (int i = 0; i < str.size(); i++) {
			if (!validateWord(str[i], banned_id[i])) {
				flag = false;
			}
		}

		if (flag) {
			vector<string> s(str.size());

			for (int i = 0; i < str.size(); i++) {
				s[i] = str[i];
			}

			ret.push_back(s);
		}
	}

	return ret;
}

int getBannedSize(vector<vector<string>>& bannedList) {
	unordered_set<string> st;

	for (vector<string> s : bannedList) {
		sort(s.begin(), s.end());

		string sb = "";
		for (int i = 0; i < s.size(); i++) {
			sb += s[i];
		}

		st.insert(sb);
	}

	return st.size();
}

int solution(vector<string> user_id, vector<string> banned_id) {
	int answer = 0;

	int listSize = banned_id.size();
	vector<bool> visit(user_id.size(), false);

	getList(0, listSize, user_id, visit);

	vector<vector<string>> bannedList = searchBannedId(list, banned_id);

	answer = getBannedSize(bannedList);

	return answer;
}

File: Programmers/java/LEVEL_3_불량_사용자.java
import java.util.*;

class Solution {
    List<String[]> list = new ArrayList<>();
    List<String> strList = new ArrayList<>();

    public int solution(String[] user_id, String[] banned_id) {
        int answer = 0;

        int listSize = banned_id.length;
        boolean[] visit = new boolean[user_id.length];

        getList(0, listSize, user_id, visit);

        List<String[]> bannedList = searchBannedId(list, banned_id);

        answer = getBannedSize(bannedList);

        return answer;
    }

    private int getBannedSize(List<String[]> bannedList) {
        Set<String> set = new HashSet<>();

        for (String[] s : bannedList) {
            Arrays.sort(s);

            StringBuilder sb = new StringBuilder();

            for (int i = 0; i < s.length; i++) {
                sb.append(s[i]);
            }

            set.add(sb.toString());
        }


        return set.size();
    }

    private List<String[]> searchBannedId(List<String[]> list, String[] banned_id) {
        List<String[]> ret = new ArrayList<>();

        for (String[] str : list) {
            boolean flag = true;
            for (int i = 0; i < str.length; i++) {
                if (!validateWord(str[i], banned_id[i])) {
                    flag = false;
                }
            }

            if (flag) {
                String[] s = new String[str.length];

                for (int i = 0; i < str.length; i++) {
                    s[i] = str[i];
                }

                ret.add(s);
            }
        }

        return ret;
    }

    private boolean validateWord(String word, String bannedWord) {
        if (word.length() != bannedWord.length()) {
            return false;
        }

        for (int i = 0; i < word.length(); i++) {
            if (bannedWord.charAt(i) == '*') {
                continue;
            }

            if (word.charAt(i) != bannedWord.charAt(i)) {
                return false;
            }
        }

        return true;
    }

    private void getList(int cnt, int limit, String[] userId, boolean[] visit) {
        if (cnt == limit) {
            String[] st = new String[strList.size()];
            for (int i = 0; i < strList.size(); i++) {
                st[i] = strList.get(i);
            }

            list.add(st);

            return;
        } else {
            for (int i = 0; i < userId.length; i++) {
                if (visit[i] == true) continue;
                visit[i] = true;
                strList.add(userId[i]);
                getList(cnt + 1, limit, userId, visit);
                visit[i] = false;
                strList.remove(strList.size() - 1);
            }
        }
    }
}